esphome/components/esp32/preferences.cpp

```diff
@@ -97,12 +100,12 @@ class ESP32Preferences : public ESPPreferences {
   ESPPreferenceObject make_preference(size_t length, uint32_t type) override {
     auto *pref = new ESP32PreferenceBackend();  // NOLINT(cppcoreguidelines-owning-memory)
     pref->nvs_handle = nvs_handle;
-    current_offset += length;
+    // current_offset += length;                     // not needed
 
-    uint32_t keyval = current_offset ^ type;
+    uint32_t keyval = type;  // removed current_offset from calculation
     char keybuf[16];
-    snprintf(keybuf, sizeof(keybuf), "%d", keyval);
-    pref->key = keybuf;  // copied to std::string
+    snprintf(keybuf, sizeof(keybuf), "%u", keyval);  // switched to unsigned to avoid negative numbers in keys
+    pref->key = keybuf;                              // copied to std::string
```

Comment:
Nothing wrong with your code.
shorter variant: 
```suggestion
pref->key = str_sprintf("%u", keyval);
```